packages/oats-runtime/src/redirect.ts

```diff
@@ -0,0 +1,72 @@
+import type { Response } from './server';
+import encodeUrl = require('encodeurl');
+import escapeHtml = require('escape-html');
+
+const REDIRECT_STATUSES = [300, 301, 302, 303, 305, 307, 308] as const;
+const DEFAULT_REDIRECT_STATUS = 302;
+const TEXT_HTML_CONTENT_TYPE = 'text/html';
+const TEXT_PLAIN_CONTENT_TYPE = 'text/plain';
+
+export type RedirectStatus = typeof REDIRECT_STATUSES[number];
+
+type DefaultValue<ContentType extends string> = ContentType extends
+  | typeof TEXT_HTML_CONTENT_TYPE
+  | typeof TEXT_PLAIN_CONTENT_TYPE
+  | `${typeof TEXT_HTML_CONTENT_TYPE};${string}`
+  | `${typeof TEXT_PLAIN_CONTENT_TYPE};${string}`
+  ? string
+  : null;
+
+export function redirect<
+  Status extends RedirectStatus = typeof DEFAULT_REDIRECT_STATUS,
+  ContentType extends string = typeof TEXT_HTML_CONTENT_TYPE,
+  Value = DefaultValue<ContentType>
+>(
+  url: string,
+  options: {
+    status?: Status;
+    contentType?: ContentType;
+    value?: Value;
+  } = {}
+): Response<Status, ContentType, Value, Record<string, any>> {
+  const { status = DEFAULT_REDIRECT_STATUS as Status } = options;
+  const encodedUrl = encodeUrl(url);
+
+  if (!REDIRECT_STATUSES.includes(status)) {
+    throw new Error(`Status "${status}" is not a redirect status.`);
+  }
+  const {
+    contentType = TEXT_HTML_CONTENT_TYPE as ContentType,
+    value = getDefaultRedirectValue({ encodedUrl, contentType }) as unknown as Value
+  } = options;
+
+  return {
+    status,
+    value: { contentType, value },
+    headers: {
+      Location: encodedUrl
```

Comment:
does this need to be encoded? to me it sounds like a task for the adapter if that is needed

Reply:
the way I see we should do it anyways so why do it in every adapter? besides in adapter it'll quite messy, as we need to modify Location header (and it's not even case sensitive).

Reply:
my hunch is that eg koa does the right thing already for headers

Reply:
well koa does encode url in their redirect - https://github.com/koajs/koa/blob/master/lib/response.js#L270
so does express - https://github.com/expressjs/express/blob/947b6b7d57939d1a3b33ce008765f9aba3eb6f70/lib/response.js#L893

Reply:
if you set a location or any other header for koa does it need encoding?

Reply:
> any other header for koa does it need encoding

idk maybe some encoding happens for all headers, but is it URI encoding? i can of course try to break stuff by setting non-encoded url and see what happens

Reply:
okay, I broke it by having Cyrillic in the url 😅 works with encodeURI tho